server/sonar-server/src/main/java/org/sonar/ce/queue/CeQueueImpl.java

```diff
@@ -70,61 +75,122 @@ public CeQueueImpl(DbClient dbClient, UuidFactory uuidFactory, DefaultOrganizati
     if (submissions.isEmpty()) {
       return Collections.emptyList();
     }
 
     try (DbSession dbSession = dbClient.openSession(true)) {
-      List<CeQueueDto> ceQueueDtos = from(submissions)
-        .transform(new CeTaskSubmitToInsertedCeQueueDto(dbSession, dbClient))
-        .toList();
+      List<CeQueueDto> ceQueueDtos = submissions.stream()
+        .filter(filterBySubmitOptions(options, submissions, dbSession))
+        .map(submission -> addToQueueInDb(dbSession, submission))
+        .collect(Collectors.toList());
       List<CeTask> tasks = loadTasks(dbSession, ceQueueDtos);
       dbSession.commit();
       return tasks;
     }
   }
 
+  private Predicate<CeTaskSubmit> filterBySubmitOptions(SubmitOption[] options, Collection<CeTaskSubmit> submissions, DbSession dbSession) {
+    EnumSet<SubmitOption> submitOptions = toSet(options);
+
+    if (submitOptions.contains(UNIQUE_QUEUE_PER_COMPONENT)) {
+      Set<String> componentUuids = submissions.stream()
+        .map(CeTaskSubmit::getComponentUuid)
+        .filter(Objects::nonNull)
+        .collect(MoreCollectors.toSet(submissions.size()));
+      if (componentUuids.isEmpty()) {
+        return t -> true;
+      }
+      return new NoPendingTaskFilter(dbSession, componentUuids);
+    }
+
+    return t -> true;
+  }
+
+  private class NoPendingTaskFilter implements Predicate<CeTaskSubmit> {
+    private final Map<String, Integer> queuedItemsByComponentUuid;
+
+    private NoPendingTaskFilter(DbSession dbSession, Set<String> componentUuids) {
+      queuedItemsByComponentUuid = dbClient.ceQueueDao().countByStatusAndComponentUuids(dbSession, PENDING, componentUuids);
+    }
+
+    @Override
+    public boolean test(CeTaskSubmit ceTaskSubmit) {
+      String componentUuid = ceTaskSubmit.getComponentUuid();
+      return componentUuid == null || queuedItemsByComponentUuid.getOrDefault(componentUuid, 0) == 0;
+    }
+  }
+
+  private static EnumSet<SubmitOption> toSet(SubmitOption[] options) {
+    return Arrays.stream(options).collect(toEnumSet(SubmitOption.class));
+  }
+
+  private CeQueueDto addToQueueInDb(DbSession dbSession, CeTaskSubmit submission) {
+    CeQueueDto dto = new CeQueueDto();
+    dto.setUuid(submission.getUuid());
+    dto.setTaskType(submission.getType());
+    dto.setComponentUuid(submission.getComponentUuid());
+    dto.setStatus(PENDING);
+    dto.setSubmitterLogin(submission.getSubmitterLogin());
+    dto.setStartedAt(null);
+    dbClient.ceQueueDao().insert(dbSession, dto);
+    return dto;
+  }
+
   protected CeTask loadTask(DbSession dbSession, CeQueueDto dto) {
     if (dto.getComponentUuid() == null) {
       return new CeQueueDtoToCeTask(defaultOrganizationProvider.get().getUuid()).apply(dto);
     }
-    com.google.common.base.Optional<ComponentDto> componentDto = dbClient.componentDao().selectByUuid(dbSession, dto.getComponentUuid());
+    Optional<ComponentDto> componentDto = dbClient.componentDao().selectByUuid(dbSession, dto.getComponentUuid());
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Annotate the parameter with @javax.annotation.Nullable in method 'selectByUuid' declaration, or make sure that null can not be passed as argument. ([squid:S4449](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS4449))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&pullRequest=3113&issues=AWIgcKDgoomSmF0fSIAW&open=AWIgcKDgoomSmF0fSIAW)

[ISSUE_KEY]: AWIgcKDgoomSmF0fSIAW